File: algorithm_datastructure/leetcode/637.average-of-levels-in-binary-tree.py
#
# @lc app=leetcode id=637 lang=python3
#
# [637] Average of Levels in Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

import collections


class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        # Use BFS, the queue is used to replace the recursion
        q = collections.deque()
        ret = []
        q.append(root)
        while q:
            size = len(q)
            row = []
            # Iterate over the current layer
            for _ in range(size):
                node = q.popleft()
                if not node:
                    continue
                # Preorder visit
                row.append(node.val)  # current value
                # Recursive for the left tree and right tree
                q.append(node.left)
                q.append(node.right)
            if row:
                ret.append(sum(row) / len(row))

        return ret


# @lc code=end

File: algorithm_data_structure/leetcode/637.average-of-levels-in-binary-tree.cpp
/*
 * @lc app=leetcode id=637 lang=cpp
 *
 * [637] Average of Levels in Binary Tree
 */

// @lc code=start

#include <algorithm>
#include <queue>
#include <vector>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
   public:
	vector<double> averageOfLevels(TreeNode *root) {
		vector<double> ans;
		if (not root) {
			return ans;
		}

		queue<TreeNode *> q;
		q.push(root);

		while (not q.empty()) {
			int count = q.size();  // Number of nodes in current layer
			double sum = 0.0;
			// Iterate over current layer
			for (int i = 0; i < count; i++) {
				TreeNode *node = q.front();
				q.pop();
				sum = sum + node->val;

				if (node->left) {
					q.push(node->left);
				}
				if (node->right) {
					q.push(node->right);
				}
			}

			ans.push_back(sum / count);
		}

		return ans;
	}
};
// @lc code=end
